File: Java/Silver/2178.java
import java.util.*;
import java.io.*;


public class Main {

    private static char[][] board;
    private static int[][] visited;

    private static int N, M;
    private static int[] moveY = new int[] {-1, 0, 1, 0};
    private static int[] moveX = new int[] {0, -1, 0, 1};

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    public static void main(String[] agrs) throws IOException {

        settingInput();

        bfs(new int[]{0,0});
        System.out.println(visited[N-1][M-1]);
    }

    private static void settingInput() throws IOException {
        int[] NM = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        N = NM[0]; M = NM[1];
        board = new char[N][M];
        visited = new int[N][M];
        for(int i=0;i<N;i++) board[i] = br.readLine().toCharArray();
    }

    private static void bfs(int[] coord) { // coord: {y,x}
        Queue<int[]> queue = new ArrayDeque<>();
        queue.offer(coord);
        visited[coord[0]][coord[1]] = 1;
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            for(int i=0;i<4;i++) {
                int nextY = current[0] + moveY[i];
                int nextX = current[1] + moveX[i];
                if (!isOutOfBoundary(nextY, nextX) && isMovable(nextY, nextX) && visited[nextY][nextX] == 0) {
                    queue.offer(new int[] {nextY, nextX});
                    visited[nextY][nextX] = visited[current[0]][current[1]] +1;
                }
            }
        }
    }

    private static boolean isOutOfBoundary(int y, int x) {
        return y<0 || x<0 || y>=N || x>=M;
    }

    private static boolean isMovable(int y, int x) {
        return board[y][x] == '1';
    }

}

File: Kotlin/Silver/2178.kt
import java.io.*;
import java.util.*;
import kotlin.collections.ArrayDeque
import kotlin.collections.ArrayList

var N = 0
var M = 0
lateinit var board: ArrayList<String>
lateinit var visited: ArrayList<IntArray>
val moveY = intArrayOf(-1, 0, 1, 0)
val moveX = intArrayOf(0, -1, 0, 1)


fun main() {

    settingInput()
    bfs(0,0)
    println(visited[N-1][M-1])

}

fun settingInput() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val NM = br.readLine().split(" ").map { it.toInt() }
    N = NM[0]
    M = NM[1]
    board = ArrayList()
    repeat(N) { board.add(br.readLine()) }
    visited = ArrayList()
    repeat(N) { visited.add(IntArray(M)) }
}

fun bfs(y: Int, x: Int) {
    val queue = ArrayDeque<Pair<Int,Int>>() // y, x
    queue.addLast(Pair(y,x))
    visited[y][x] = 1
    while (queue.isNotEmpty()) {
        val itm = queue.removeFirst()
        for (i in 0..3) {
            val nextY = itm.first + moveY[i]
            val nextX = itm.second + moveX[i]
            if (!isOutOfBoundary(nextY, nextX) && isMovable(nextY, nextX) && visited[nextY][nextX] == 0) {
                queue.addLast(Pair(nextY, nextX))
                visited[nextY][nextX] = visited[itm.first][itm.second] +1
            }

        }
    }

}

fun isOutOfBoundary(y: Int, x: Int): Boolean {
    return y<0 || x<0 || y>=N || x>=M
}

fun isMovable(y: Int, x: Int): Boolean {
    return board[y][x] == '1'
}